### PR TITLE
feat: Cognito Advance Security features

### PR DESCRIPTION
**Pull request by elorzafe**
#### Description of changes

Add support for Cognito Advance Security features on Cognito APIs that this is supported:
- `confirmResetPassword` ([Cognito ConfirmForgotPassword API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmForgotPassword.html))
- `confirmSignIn` ([Cognito RespondToAuthChallenge API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_RespondToAuthChallenge.html))
- `confirmResetPassword` ([Cognito ConfirmForgotPassword API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmForgotPassword.html))
- `confirmSignUp`([Cognito ConfirmSignUp API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmSignUp.html))
- `resendSignUpCode` ([Cognito ResendConfirmationCode API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ResendConfirmationCode.html))
- `resetPassword` ([Cognito ForgotPassword API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ForgotPassword.html))
- `fetchAuthSession` when refresh token is required ([Cognito InitiateAuth API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html))
-`signIn` on all different Auth modes ([Cognito InitiateAuth API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html))

#### Description of how you validated changes
Unit tests
run sample app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/apis/confirmSignUp.ts [hunk 1]
```diff
@@ -39,19 +44,31 @@ export async function confirmSignUp(
 		AuthValidationErrorCode.EmptyConfirmSignUpCode
 	);
 
+	const jsonReq = {
+		Username: username,
+		ConfirmationCode: confirmationCode,
+		ClientMetadata: clientMetadata,
+		ForceAliasCreation: options?.serviceOptions?.forceAliasCreation,
+		ClientId: authConfig.userPoolClientId,
+		// TODO: handle UserContextData
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmSignUp.ts:53]:
```suggestion
```

File: packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts
Comment (jimblanc) [inline packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts:1]: File naming nit: `userContextData.native.ts` & `userContextData.ts` to make it clear what this is providing.

File: packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// TODO: add support after https://amazon-cognito-assets.us-east-1.amazoncognito.com/amazon-cognito-advanced-security-data.min.js can be imported
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts:1]: Just to confirm, no loss in functionality here right? We don't support this in RN currently?

**Review by jimblanc — COMMENTED** [review]
Looks pretty good, some nits

File: packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts (hunk 2 above)
Comment (elorzafe) [reply packages/auth/src/providers/cognito/utils/advanceSecurity.native.ts:1]: correct, there is an open feature request

File: packages/auth/src/providers/cognito/apis/confirmResetPassword.ts [hunk 3]
```diff
@@ -42,18 +46,29 @@ export async function confirmResetPassword(
...
+	};
+
+	const UserContextData = getUserContextData({
+		username,
+		userPoolId,
+		userPoolClientId,
+	});
+
+	if (UserContextData) {
+		(jsonReq as any)['UserContextData'] = UserContextData;
```
[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/auth/src/providers/cognito/apis/confirmResetPassword.ts:64]: I think it makes sense, I was a little hesitant to include the body with `undefined`

[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/auth/src/providers/cognito/apis/confirmResetPassword.ts]: done

File: packages/auth/src/providers/cognito/apis/confirmResetPassword.ts [hunk 4]
```diff
@@ -43,17 +47,24 @@ export async function confirmResetPassword(
...
+			region: getRegion(authConfig.userPoolId),
+			userAgentValue: getAuthUserAgentValue(AuthAction.ConfirmResetPassword),
 		},
 		{
 			Username: username,
 			ConfirmationCode: code,
 			Password: newPassword,
 			ClientMetadata: metadata,
 			ClientId: authConfig.userPoolClientId,
+			UserContextData: UserContextData,
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/confirmResetPassword.ts:67]: I believe this is a ts lint rule

```suggestion
                        UserContextData
```

**Review by israx — APPROVED** [review]
LGTM. Thanks @elorzafe

**Review by jimblanc — DISMISSED** [review]

**Review by jimblanc — APPROVED** [review]

**Review by israx — APPROVED** [review]

**Pull request merged**